Added overload to allow using custom prefix name and flat folder stru…

yogigrantz opened the pull request:
…cture

Added overload to allow using custom prefix name and flat folder structure in Azure Blob. In addition to the original only option  $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}", the overload can produce Blob output in this format :$"{appName}/{_customPrefixFileName}{currDate.Year}{currDate.Month:00}{currDate.Day:00}.txt" if custom prefix filename is specified in the log provider Dependency Injection

This is a do over of pull request #25078 with fresh Refork and Reclone from donetcore master branch as of 10/30/2020, 11:05AM PST

Summary of the changes (Less than 80 chars)
 - Added an overload for the entry point to accept customPrefix name
 - Added a condition in the main logic based on customPrefix name to output flat folder structure in Azure Blob Log

Addresses  #25885

Commit:
Added overload to allow using custom prefix name and flat folder structure

Added overload to allow using custom prefix name and flat folder structure in Azure Blob. In addition to the original only option  $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}", the overload can produce Blob output in this format :$"{appName}/{_customPrefixFileName}{currDate.Year}{currDate.Month:00}{currDate.Day:00}.txt" if custom prefix filename is specified in the log provider Dependency Injection

Commit:
Fixed Unit Test issues  and added the overload to PublicAPI registry

1. Fixed Unit Test issues: added the necessary parameter to call the modified internal method.
2. Added the overload to PublicAPI registry in Unshipped.txt

yogigrantz commented:
I'm getting this error:
.packages\microsoft.dotnet.helix.sdk\6.0.0-beta.20526.10\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets#L76 

I did add the overload to the unshipped.txt and removed .\artifacts\obj\Microsoft.Extensions.Logging.AzureAppServices\ as per @dougbu  's advice . The VS built OK, and I'm getting 22 successful checks on this one. Only 3 failure, most of them on helix sdk\6.0.0-beta.20526.10  

@wtgodbe

dougbu commented:
Failures look like an infrastructure issue @yogigrantz. I had AzDO rerun the failed jobs.

yogigrantz commented:
oh good, it built successfully now. Thank you @dougbu  

@Pilchie  @davidfowl

Pilchie commented:
@JunTaoLuo - can you take a look at this?  Also tagging @davidfowl

Commit:
Merge branch 'main' into master

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
         {
             var serviceCollection = new ServiceCollection();
-            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
+            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext, null));
```
This shouldn't be needed, the overload should set the prefix correctly. If this doesn't work, there is a problem with back compatibility.
```suggestion
            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+        }
+
+        [Fact]
+        public void BuilderExtensionAddsConfigurationChangeTokenSourceOrig()
```
```suggestion
        public void BuilderExtensionAddsConfigurationChangeTokenSource()
```
No need to add Orig at the end.

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+        }
+
+        [Fact]
+        public void BuilderExtensionAddsConfigurationChangeTokenSourceOverload()
```
```suggestion
        public void BuilderExtensionAddsConfigurationChangeTokenSourceCustomPrefix()
```

Overload isn't descriptive.

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
 
         [Fact]
-        public void BuilderExtensionAddsIConfigureOptions()
+        public void BuilderExtensionAddsIConfigureOptionsOrig()
```
```suggestion
        public void BuilderExtensionAddsIConfigureOptions()
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+            // Tracking for main configuration
+            Assert.Equal(1, serviceCollection.Count(d => d.ServiceType == typeof(IOptionsChangeTokenSource<LoggerFilterOptions>)));
+
+            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext, null));
```
```suggestion
            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
             Assert.NotEqual(0, count);
 
-            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
+            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext, null));
```
```suggestion
            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+            // Tracking for main configuration
+            Assert.Equal(2, serviceCollection.Count(d => d.ServiceType == typeof(IConfigureOptions<LoggerFilterOptions>)));
+
+            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext, null));
```
```suggestion
            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
 
         [Fact]
-        public void LoggerProviderIsResolvable()
+        public void LoggerProviderIsResolvableOrig()
```
```suggestion
        public void LoggerProviderIsResolvable()
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+        }
+
+        [Fact]
+        public void BuilderExtensionAddsIConfigureOptionsOverload()
```
```suggestion
        public void BuilderExtensionAddsIConfigureOptionsCustomPrefix()
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+        public void LoggerProviderIsResolvableOrig()
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext, null));
```
```suggestion
            serviceCollection.AddLogging(builder => builder.AddAzureWebAppDiagnostics(_appContext));
```

JunTaoLuo commented on src/Logging.AzureAppServices/test/LoggerBuilderExtensionsTests.cs:
```diff
@@ -24,53 +24,106 @@ public LoggerBuilderExtensionsTests()
+        }
+
+        [Fact]
+        public void LoggerProviderIsResolvableOverload()
```
```suggestion
        public void LoggerProviderIsResolvableCustomPrefix()
```

JunTaoLuo commented on src/Logging.AzureAppServices/src/BlobLoggerProvider.cs:
```diff
@@ -55,15 +56,25 @@ public BlobLoggerProvider(IOptionsMonitor<AzureBlobLoggerOptions> options)
+
+                if (!string.IsNullOrEmpty(_customPrefixFileName))
+                {
+                    string filename = $"{_customPrefixFileName}{currDate.Year}{currDate.Month:00}{currDate.Day:00}.txt";
```
```suggestion
                    var filename = $"{_customPrefixFileName}{currDate.Year}{currDate.Month:00}{currDate.Day:00}.txt";
```

JunTaoLuo commented on src/Logging.AzureAppServices/src/BlobLoggerProvider.cs:
```diff
@@ -55,15 +56,25 @@ public BlobLoggerProvider(IOptionsMonitor<AzureBlobLoggerOptions> options)
             foreach (var eventGroup in eventGroups)
             {
                 var key = eventGroup.Key;
-                var blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}";
```
I tracked down the original change that used the old key format in https://github.com/aspnet/Logging/pull/637. However, it doesn't explain why that format is used. My main concern is that the new API actually changes the folder structure but my concerns are lessened somewhat because it's only when the new API is used and non of the previous behaviours are changed.

This change of format depending on whether a prefix is passed feels very strange and should at least clearly documented in the xml doc comments.

yogigrantz replied:
Yes I hear you Jun Tao. Should we establish an enum that drives the output file structure? For example FileStructure.Flat for the new overload, and FileStructure.Hierarchy for the original functionality. Then the custom file prefix could be used for FileStructure.Flat only?

JunTaoLuo commented on src/Logging.AzureAppServices/src/BlobLoggerProvider.cs:
```diff
@@ -55,15 +56,25 @@ public BlobLoggerProvider(IOptionsMonitor<AzureBlobLoggerOptions> options)
+                    string filename = $"{_customPrefixFileName}{currDate.Year}{currDate.Month:00}{currDate.Day:00}.txt";
+                    blobName = $"{_appName}/{filename}";
+                }
+                else
+                    blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}";
```
```suggestion
                else
                {
                    blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}";
                }
```

nit, this is the style we use in this repo.

JunTaoLuo commented on src/Logging.AzureAppServices/src/BlobLoggerProvider.cs:
```diff
@@ -55,15 +56,25 @@ public BlobLoggerProvider(IOptionsMonitor<AzureBlobLoggerOptions> options)
         internal override async Task WriteMessagesAsync(IEnumerable<LogMessage> messages, CancellationToken cancellationToken)
         {
             var eventGroups = messages.GroupBy(GetBlobKey);
+            DateTime currDate = DateTime.Now;
```
We want to use the time from the key not the current date.

JunTaoLuo reviewed (commented):
I'm a bit hesitant to take the PR since I don't know why the old key format was used in the first place. Also, if we want the format to be customizable, why don't we expose an API that allows the user to specify the filename and directory structure instead of only being able to specify a prefix?

dougbu commented on src/Logging.AzureAppServices/src/AzureBlobLoggerOptions.cs:
```diff
@@ -30,6 +29,8 @@ public string BlobName
             }
         }
 
+        internal string CustomPrefixFileName { get; set; }
```
Suggest moving "Prefix" to the end because it seems more natural
```suggestion
        internal string CustomFileNamePrefix { get; set; }
```

Side note: Wish we were consistent about "filename" (more common) vs. "fileName" and "Filename" (less common) vs. "FileName" in this repo.

yogigrantz replied:
- Changed to CustomFileNamePrefix

Commit:
Update LoggerBuilderExtensionsTests.cs

Restored the unit test original code to test the original extension  .AddAzureWebAppDiagnostics without custom prefix by removing the null parameter for custom prefix. This will ensure backward compatibility because it prevents the unit test from calling the overload function.
Restored the name of the original unit test function
Changed the name of the overload unit test function to xxxCustomPrefix()

Commit:
Update BlobLoggerProvider.cs

Commit:
Renamed property

 Renamed from CustomPrefixFileName to CustomFileNamePrefix for better grammar

Commit:
Changed field name

changed from customPrefixFilename to customFilenamePrefix

yogigrantz reviewed (commented):
Changes made, please kindly review. Thank you

Commit:
Update AzureAppServicesLoggerFactoryExtensions.cs

Added description in XML to explain that the output will be in flat file structure

yogigrantz reviewed (commented):
Added comment to indicate that this overload will produce flat file structure

JunTaoLuo commented:
Changes look fine mechanically, but I'll need to double check the file structure format and figure out if anything else is impacted.

yogigrantz commented:
Thank you @JunTaoLuo . The ci failed on build test for macOS 10.14 and Helix x64. Is this something on my end?

JunTaoLuo commented on src/Logging.AzureAppServices/src/AzureAppServicesLoggerFactoryExtensions.cs:
```diff
@@ -21,15 +21,30 @@ public static class AzureAppServicesLoggerFactoryExtensions
+        /// <param name="builder">The extension method argument</param>
+        /// <param name="customPrefix"></param>
+        /// <returns></returns>
+        public static ILoggingBuilder AddAzureWebAppDiagnostics(this ILoggingBuilder builder, string customPrefix)
```
I spoke to some folks who were familiar with this component and it seems like changing the layout of the blob file won't be an issue. However, I'd like to suggest a different API.

It seems fairly unintuitive that using this overload would change the folder structure. Even with the API doc in place, this can be a pitfall for users this is not a pattern we have used elsewhere. Instead I propose that we add a configurable option to allow for customized blob name format. Specifically, add a formatter to `AzureBlobLoggerOptions`
```c#
Func<DateTimeOffset, string, string, string> BlobNameFormatter { get; set; }
```

By default this would use the current logic at https://github.com/dotnet/aspnetcore/blob/main/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L66 passing in the `_appName`, `_fileName` and a `DateTimeOffset` constructed with the key. To make this more user friendly, we should add an overload that takes a delegate to configure the option in the same style as https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/HttpsPolicy/src/HttpsRedirectionServicesExtensions.cs#L21 so folks can override the default logic

Could you give this a try before I bring this PR to our weekly API review?

yogigrantz replied:
Hi JunTao, I see what you're saying and totally agree that this is counter intuitive. OK I will look into using delegate like you said. I'd probably start fresh

JunTaoLuo replied:
FYI I brought this up in API review and some small changes were suggested, primarily moving the arguments to a context struct so it's more clear what the arguments are for. See the comment at https://github.com/dotnet/aspnetcore/pull/27383#issuecomment-818007583.

pranavkm commented:
API review suggestion

```diff
public class AzureBlobLoggerOptions
{
+   public Func<AzureBlobLoggerContext, string> FileNameFormat { get; set; }
}

+ public readonly struct AzureBlobLoggerContext
+ {
+     public AzureBlobLoggerContext(string appName, string identifier, DateTimeOfset timestamp);
+     public string AppName { get; } 
+     public string Identifier { get; } 
+     public DateTimeOffset Timestamp { get; }
+ }
```

yogigrantz commented:
@pranavkm  : I was trying to implement your suggestion. However the declaration of public AzureBlobLoggerContext(string appName, string identifier, DateTimeOfset timestamp); would caused a CS0501 compiler error on .net Core 3.1

Secondly, how would this setup help deciding whether to construct 
    blobName = $"{_appName}/{filename}"; 
or
   blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}"; 

We need some kind of indicator to decide whether to build a blob in a flat structure or year/month/day/hour nested structure

JunTaoLuo commented:
Hey @yogigrantz the idea is that the user can customize the logic for computing filename by overriding `FileNameFormat` of the `AzureBlobLoggerOptions`. For this to work, you'll also need to add an overload to `AddAzureWebAppDiagnostics` that takes in an `Action<AzureBlobLoggerOptions>` as a parameter and pass that to the construction of `BlobLoggerConfigureOptions` where it's used in its `Configure` method. 

With these changes, you'll be able to customize the logic by:
```c#
builder.AddAzureWebAppDiagnostics(azureBlobLoggerOptions => 
{
    azureBlobLoggerOptions.FileNameFormat = context => 
    {
        return $"{context.AppName}/customPrefix{context.Identifier}{context.Timestamp.ToString(CultureInfo.InvariantCulture)}"
    }
});
``` 

As for the CS0501 error, can you push your code to this PR so I can figure out how you are getting it?

Commit:
Added AzureBlobLoggerContext

Added AzureBlobLoggerContext as per discussion in https://github.com/dotnet/aspnetcore/pull/27383#issuecomment-818007583

yogigrantz commented:
@JunTaoLuo : I just pushed the new struct AzureBlobLoggerContext for you to see. Thank you

JunTaoLuo commented on src/Logging.AzureAppServices/src/AzureBlobLoggerContext.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public readonly struct AzureBlobLoggerContext
+    {
+        public AzureBlobLoggerContext(string appName, string identifier, DateTimeOfset timestamp);
```
Ah, well you need to define the constructor like this:
```suggestion
        public AzureBlobLoggerContext(string appName, string identifier, DateTimeOffset timestamp)
        {
            AppName = appName;
            Identifier = identifier;
            Timestamp = timestamp;
        }
```

yogigrantz replied:
Hi JunTao, the body in the constructor will now compile. But now how do we wire the new struct and the new Func with the existing API such a way that existing use case is not required to define the File format, yet the new use case can define a flat structure shown in your example. Would we still need to make changes in BlobLoggerProvider? After all, it is internal override async Task WriteMessagesAsync in BlobLoggerProvider that originally forces the folder structure to be in deeply nested folder format of $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}".

If we change BlobLoggerProvider to use the option context, then the existing applications would break because none of them are using the option context up to today.

JunTaoLuo replied:
A couple of implementation notes, let me know if these answers your question:
1. Set the default value of the `FileNameFormat` property on the options class to the logic at https://github.com/dotnet/aspnetcore/blob/9819c08be913f918daf130e85ac6fd8e3aad46c9/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L66. i.e. something like:
```c# 
public class AzureBlobLoggerOptions
{
...
   public Func<AzureBlobLoggerContext, string> FileNameFormat { get; set; } = context =>
      $"{context.AppName}/{context.Timestamp.Year}/{context.Timestamp.Month:00}/{context.Timestamp.context.Timestamp:00}/{context.Timestamp.Hour:00}/{context.Identifier}";
...
}
```
2. We store the option monitor on the class as a private member: https://github.com/dotnet/aspnetcore/blob/9819c08be913f918daf130e85ac6fd8e3aad46c9/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L53
3. The logic here: https://github.com/dotnet/aspnetcore/blob/9819c08be913f918daf130e85ac6fd8e3aad46c9/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L66 will be modified to something like:
```c#
var blobName = _options.CurrentValue.FileNameFormat(new AzureBlobLoggerContext(_appName, _fileName, new DateTimeOffset (key.Year, key.Month, key.Day, key.Hour, 0, 0, TimeSpan.Zero)));
```

This way, the current behaviour is preserved and users can use the options to override the format of the file produced to anything they want, including a flat structure. See my usage example in https://github.com/dotnet/aspnetcore/pull/27383#issuecomment-822799282. Note, an overload of `AddAzureWebAppDiagnostics` that takes the action delegate needs to be added.

Commit:
Added suggested codes

Added Func delegate and AzureBlobLoggerContext struct

JunTaoLuo commented:
I don't think I can push to your branch so I made some updates and opened a PR: https://github.com/dotnet/aspnetcore/pull/32190. Let's continue the conversation there.

yogigrantz commented:
@JunTao, my apology for not being able to respond sooner as I have company assignments that need to be taken care of. But I really appreciate your time for adding the file format options and will definitely follow along on the new PR
